[동적 계획법] 10월 4일

jaeseo222 opened the pull request:
아직 시뮬레이션이나,, 동적 계획법에 익숙해지지 못해서 많이 찾아보며 작성하였습니다,,!ㅜㅜㅜㅜㅜㅜ그래도 열심히 하였습니다....!!! 항상 좋은 피드백 감사합니닷!

Commit:
[동적 계획법] 10월 4일

아직 시뮬레이션이나,, 동적 계획법에 익숙해지지 못해서 많이 찾아보며 작성하였습니다,,!ㅜㅜㅜㅜㅜㅜ그래도 열심히 하였습니다....!!! 항상 좋은 피드백 감사합니닷!

Commit:
[동적 계획법] 10월 4일 - Update

다시 보니 잘못 쓴 부분이 있어 수정합니당

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/10844.cpp:
```diff
@@ -0,0 +1,42 @@
+int num(int n) {
+	vector<vector<int>> dp(n+1, vector<int>(10,0));
+	
+	// n=2 : 10,21,12,32,,23,43,34,,,
+	// 마지막 자리가 3으로 끝난다면 3앞에 올 수 있는 숫자는 3-1, 3+1
+	// n-1의 자리수를 가지면서 2와 4로 끝나는 숫자의 경우를 가지고 온다.
+	//i의 길이를 가지며 j로 끝나는 수의 경우의 수
+	// ex.dp[2][2]: 2 길이를 가지며 2로 끝나는 수의 경우의 수 => 12, 32
+	// 그래서 길이가 1이며 1과 3으로 끝날 때의 경우의 수를 구하면 된다. = dp[1][1]+dp[1][3]
```
👍👍

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/10844.cpp:
```diff
@@ -0,0 +1,42 @@
+﻿#include<iostream>
+#include<vector>
+
+#define mod 1000000000
```
상수는 const로 선언해주세요! 그리고 상수는 어떻게 표기하기로 했었나요?

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/1149.cpp:
```diff
@@ -0,0 +1,36 @@
+	//최소비용
+	int tmp = dp[n][0];
+	for (int i = 1; i < 3; i++) {
+		if (tmp > dp[n][i]) tmp = dp[n][i];
+	}
```
여기는 if문을 사용하지 않으면 어떻게 구현할 수 있나요?
이럴 때에 유용한 함수가 하나 있었어요!

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/17404.cpp:
```diff
@@ -0,0 +1,49 @@
+
+	int ans = 987654321;//큰 수
+	for (int j = 0; j < 3; j++) {
+		//첫 번째 집은 각각의 색으로 시작
+		dp[0][0] = ans;
+		dp[0][1] = ans;
+		dp[0][2] = ans;
+		// 이번에 칠할 색상(첫 번째 집에 칠해져 있는 색 제외한 색)
+		dp[0][j] = cost[0][j];
+
+		for (int i = 1; i < n; i++) {
+			// 빨강으로 칠해지기 위해서, 전이 초록, 파랑
+			// 초록으로 칠해지기 위해서, 전이 빨강, 파랑
+			// 파랑으로 칠해지기 위해서, 전이 빨강, 초록
+			dp[i][0] = min(dp[i - 1][1], dp[i - 1][2]) + cost[i][0];
+			dp[i][1] = min(dp[i - 1][0], dp[i - 1][2]) + cost[i][1];
+			dp[i][2] = min(dp[i - 1][0], dp[i - 1][1]) + cost[i][2];
+		}
+
+		//첫 번째 집 색깔이 j인 경우, 마지막 집은 j가 아닌 색깔로 칠함
+		for (int i = 0; i < 3; i++)
+			if (i != j) ans = min(ans, dp[n - 1][i]);//최솟값
```
이 부분은 1149번과 거의 같아요!
그럼 함수로 빼면 어떨까요?

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/20923.cpp:
```diff
@@ -0,0 +1,66 @@
+﻿#include<iostream>
+#include<queue>
+using namespace std;
+
+deque<int> deck[2], ground[2];//0은 도도꺼, 1은 수연꺼
```
👍

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/20923.cpp:
```diff
@@ -0,0 +1,66 @@
+
+int bellWin() {
+	//그라운드가 어느 쪽도 비어있으면 안된다.
+	//그라운드에 있는 카드 더미에서 가장 위에 위치한 카드의 숫자가 5 : 도도 종침
+	if (ground[0].size() && ground[0].front() == 5 || ground[1].size() && ground[1].front() == 5) return 0;
```
비어있는 여부는 size보단 empty로 체크하는게 좋아요!

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/20923.cpp:
```diff
@@ -0,0 +1,66 @@
+		//4. 카드 가져오기
+		if (w != -1)
+			mergeCard(w);
+
+		turn = 1 - turn; //상대방 차례
```
turn은 도도거나/도도가 아니거나로 표현할 수도 있어요! 그럼 자료형이 어떻게 되나요?

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/20923.cpp:
```diff
@@ -0,0 +1,66 @@
+
+		turn = 1 - turn; //상대방 차례
+	}
+}
+int main() {
```
잘 구현하셨네요!
특히 덱 배열을 사용하셔서 중복 코드를 많이 줄이고 함수를 깔끔하게 만드신게 정말 좋아요!!

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/3190.cpp:
```diff
@@ -0,0 +1,109 @@
+#include<queue>
+using namespace std;
+const int SIZE = 101;
+struct Pos {//위치 정보
+	int y, x;
+};
+struct Dir {//방향 전환 정보
+	int t;// t초가 끝난 뒤에
+	char c;//왼쪽:L 오른쪽: D 으로 90도 회전한다.
+};
+struct Joint {//뱀의 꺾인 부분
+	int y, x, d;
+};
```
구조체도 그냥 소문자로 선언하는게 좋아요!
자료형과 같은 의미로 사용되는데 cpp에서 자료형은 소문자로 쓰니까요!

HyunJiLim0406 commented on 9월 28일 - 동적 계획법/3190.cpp:
```diff
@@ -0,0 +1,109 @@
+			direction.pop();
+		}
+
+		//꼬리가 뱀의 꺾인 부분에 도달했을 시, 꼬리 방향 전환
+		if (!joint.empty() && tail.y == joint.front().y && tail.x == joint.front().x) {
+			tail_dir = joint.front().d;
+			joint.pop();
+		}
```
p2
머리와 꼬리를 따로 저장해서 꼬리를 따로 조작해줘야 하는 아쉬움이 있어요!
뱀 자체를 어딘가에 저장하면 어떨까요? 머리가 늘어나고, 꼬리가 줄어든다는건 다시 말하면 머리에서 삽입이, 꼬리에서 삭제가 일어난다고 할 수도 있겠네요?

HyunJiLim0406 reviewed (approved):
p2
수고하셨습니다! 모르는게 있을 때 찾아보고 공부하는건 좋은거예요! 저도 아직 모르는 문제가 있으면 많이 검색해봐요~ 검색을 하면 타인의 코드를 해석하는 능력도 키울 수 있구요!! 그냥 코드를 베끼기 위한 검색이 아니라 공부를 위한 검색은 저희도 권장합니다~ 핳 저희 구글링학과잖아요

코드는 다 잘 작성해주셨어요! 특히 20923번은 정말 잘해주셨네요~!!
근데 3190번은 구현 방향을 아주 살짝만 바꿔보는거 어떨까요? 힌트는 코드에 달아놓았어요!
그리고 dp 연산을 하는 부분은 다 함수로 빼서 메인을 최대한 짧게 해주세요!
다른 코멘트들도 참고해서 수정해주시고 리뷰어로 저 호출해주세요~

앗 그리고 저희 샘플코드들도 올라오면 많이 참고해주세요! 구글링 1페이지에 나오는 블로그들보다는 잘 쓰려고 항상 노력하는 중이랍니다 하하핳

Commit:
[동적 계획법] 10월 5일 - Update

먼저 피드백 해 주신 deque방법이 쉽게 떠오르지 않아 튜터님께서 올려주신 풀이에 주석을 달며 공부하였습니다! 뱀 자체를 deque로 구현해주신 것을 보고 pop, push할 때의 순서나 조건이 중요한 것을 상기시키며 이해하였습니다..! 항상 감사합니다!!!!

JiyooonPark reviewed (commented):
튜티님! 혹시 3190번에 대한 추가 제출을 하시는것이 맞을까요? 추가제풀과 정규제출이 같은 PR에 있어서 확인이 어렵네요..! 다음부터는 정규 제출과 추가 제출은 별도의 PR로 나눠주시면 감사하겠습니다!

jaeseo222 commented:
아하!!! 확인하기 불편하셨겠군요,,!!! 넵 알겠습니다!! 다음부터는 꼭 주의하겠습니다. 감사합니다!!!! 혹시 추가제출에서 부족한 점이 있을까요?!
